src/app/inventory/store/sockets.ts

```diff
@@ -409,6 +416,92 @@ function addPlugOption(
+  SocketCategoryHashes.ArmorPerks_LargePerk,
+  SocketCategoryHashes.ArmorPerks_Reusable,
+  SocketCategoryHashes.ArmorTier,
+  SocketCategoryHashes.ClanPerks_Unlockable_ClanBanner,
+  SocketCategoryHashes.GhostShellPerks,
+  SocketCategoryHashes.VehiclePerks,
+];
+/**
+ * DIM sometimes wants to know whether a plug is the "empty" plug so that
+ * it knows not to record an override, or it may choose to reset a socket
+ * back to empty to free up mod space, or it may wish to distinguish the
+ * empty plug in UI sorting. However there's no easy, manifest-driven way
+ * to figure out whether an empty plug exists and if so, what it is.
+ *
+ * The closest thing is the singleInitialItemHash, and that works for many
+ * mod sockets, but it's insufficient in some cases (non-exhaustive):
+ *
+ * 1. The socket may not have a singleInitialItemHash. Artifice artifact mod
+ *    slots don't reference any plug in singleInitialItemHash, and the first
+ *    entry in the plug set just so happened to be the empty plug.
+ * 2. The socket's singleInitialItemHash may reference a non-empty plug. A lot
+ *    of armor references associated shaders here instead.
+ * 3. The singleInitialItemHash is an empty plug, but it's not the proper empty
+ *    plug. This happens to void subclass aspect and fragment sockets, and is
+ *    really insidious because void subclasses start with these sockets but these
+ *    plugs can never be inserted, so we can't use it.
+ */
+function findEmptyPlug(
+  socket: DestinyItemSocketEntryDefinition,
+  socketType: DestinySocketTypeDefinition,
+  plugSet: DimPlugSet | undefined,
+  reusablePlugs?: DestinyItemPlugBase[]
+) {
+  // First, perform some filtering so we don't repeatedly search through very large plug sets
+  // like armor 2.0 stat plug sets, masterwork sets, or armor energy sets.
+
+  if (noDefaultSocketCategoryHashes.includes(socketType.socketCategoryHash)) {
+    return undefined;
+  }
+
+  // Y2+ weapon masterworks don't have an "empty" entry.
+  if (socket.socketTypeHash === weaponMasterworkY2SocketTypeHash) {
+    return undefined;
+  }
+  // Exotic mods (like the Aeon socket) can't be emptied.
+  if (
+    socketType.plugWhitelist.length &&
+    socketType.plugWhitelist.every((e) => e.categoryIdentifier.includes('enhancements.exotic'))
+  ) {
+    return undefined;
+  }
+  // Sockets that ONLY get their items from your inventory necessarily can't be emptied
+  if ((socket.plugSources & ~SocketPlugSources.InventorySourced) === 0) {
+    return undefined;
+  }
+
+  // Check sources in decreasing order of assumed reliability:
+  // When the live API response tells us about a certain empty plug, that's
+  // probably the most correct. PlugSets are usually better than the
+  // socket.reusablePlugs, but sometimes there's an empty option not present in the PlugSet.
+  // FIXME #7793: reusablePlugItems is thrown away when there's a PlugSet.
+  const isDefault = (p: number) => emptyPlugHashes.has(p);
```

Comment:
not sure why this function needs to be scoped to just here and not just once at top level.
does it not get re-allocated each findEmptyPlug call? maybe v8 is smarter than that, but :shrug: